src/storage/segcache/src/segcache.rs

```diff
@@ -0,0 +1,288 @@
+        self.hashtable.get_no_freq_incr(key, &mut self.segments)
+    }
+
+    /// Insert a new item into the cache. May return an error indicating that
+    /// the insert was not successful.
+    /// ```
+    /// use segcache::{CoarseDuration, Policy, SegCache};
+    ///
+    /// let mut cache = SegCache::builder().build();
+    /// assert!(cache.get(b"drink").is_none());
+    ///
+    /// cache.insert(b"drink", b"coffee", None, CoarseDuration::ZERO);
+    /// let item = cache.get(b"drink").expect("didn't get item back");
+    /// assert_eq!(item.value(), b"coffee");
+    ///
+    /// cache.insert(b"drink", b"whisky", None, CoarseDuration::ZERO);
+    /// let item = cache.get(b"drink").expect("didn't get item back");
+    /// assert_eq!(item.value(), b"whisky");
+    /// ```
+    pub fn insert<'a>(
+        &mut self,
+        key: &'a [u8],
+        value: &[u8],
+        optional: Option<&[u8]>,
+        ttl: CoarseDuration,
+    ) -> Result<(), SegCacheError<'a>> {
+        // default optional data is empty
+        let optional = optional.unwrap_or(&[]);
+
+        // calculate size for item
+        let size = (((ITEM_HDR_SIZE + key.len() + value.len() + optional.len()) >> 3) + 1) << 3;
+
+        // try to get a `ReservedItem`
+        let mut retries = RESERVE_RETRIES;
+        let reserved;
+        loop {
+            match self
+                .ttl_buckets
+                .get_mut_bucket(ttl)
+                .reserve(size, &mut self.segments)
+            {
+                Ok(mut reserved_item) => {
+                    reserved_item.define(key, value, optional);
+                    reserved = reserved_item;
+                    break;
+                }
+                Err(TtlBucketsError::ItemOversized { size }) => {
+                    return Err(SegCacheError::ItemOversized { size, key });
+                }
+                Err(TtlBucketsError::NoFreeSegments) => {
+                    if self
+                        .segments
+                        .evict(&mut self.ttl_buckets, &mut self.hashtable)
+                        .is_err()
+                    {
+                        retries -= 1;
+                    } else {
+                        continue;
+                    }
+                }
+            }
+            if retries == 0 {
+                return Err(SegCacheError::NoFreeSegments);
+            }
+            retries -= 1;
+        }
+
+        // insert into the hashtable, or roll-back by removing the item
+        // TODO(bmartin): we can probably roll-back the offset and re-use the
+        // space in the segment, currently we consume the space even if the
+        // hashtable is overfull
+        if self
+            .hashtable
+            .insert(
+                reserved.item(),
+                reserved.seg(),
+                reserved.offset() as u64,
+                &mut self.ttl_buckets,
+                &mut self.segments,
+            )
+            .is_err()
```

Comment:
somewhat random question- is there community agreement on whether exception should be generally listed before normal case (C's convention afaict) or after (Python/Java conventions) for Rust code?